tests/test_grad_accum.py

```diff
@@ -0,0 +1,149 @@
+import torch
+from torch import optim
+import copy
+
+from open_lm.train import train_one_epoch
+from open_lm.main import random_seed
+from open_lm.model import create_model
+from open_lm.data import get_data
+from open_lm.scheduler import cosine_lr
+from tests.shared import MockArgs
+
+
+def create_optimizer(args, model):
+    named_parameters = list(model.named_parameters())
+    params = [p for _, p in named_parameters if p.requires_grad]
+    optimizer = optim.AdamW(
+        [
+            {"params": params, "weight_decay": args.wd},
+        ],
+        lr=args.lr,
+        betas=(args.beta1, args.beta2),
+        eps=args.eps,
+    )
+    return(optimizer)
+
+
+def test_grad_acc(accum_freq = 2, threshold = 1e-7):
+    args = MockArgs("open_lm_11m")
+    
+    # only want to look at one batch
+    args.train_num_samples = args.batch_size
+    
+    # increase learning rate and remove warmup for maximize change to model weights 
+    args.lr = 2
+    args.warmup = 0
+    
+    # create models
+    random_seed()
+    model_accum_grad = create_model(args).to(args.device)
+    model_no_accum_grad = copy.deepcopy(model_accum_grad).to(args.device) #should I be using copy.deepcopy or clone? or model2.load_state_dict(model1.state_dict())
+
+    # check that models weights are similar (within some threshold)
+    for p1, p2 in zip(model_accum_grad.parameters(), model_no_accum_grad.parameters()):
+        assert torch.allclose(p1, p2, atol=threshold)
+    
+    # create dataloader
+    data = get_data(
+        args,
+        epoch=0,
+        tokenizer=None,
+        skip_train=False,
+    )
+
+    # create optimizer
+    optimizer = create_optimizer(args, model_accum_grad)
+    
+    # create scheduler
+    scheduler = cosine_lr(
+        optimizer,
+        args.lr,
+        args.warmup,
+        10,
+        args.lr_cooldown_end,
+        args.force_min_lr,
+    )
+
+    # create loss
+    loss = torch.nn.CrossEntropyLoss()
+
+    # train on mock data with/without grad accumulation for one epoch
+    for model, accum_freq in zip([model_accum_grad, model_no_accum_grad], [accum_freq, 1]):
+        args.accum_freq = accum_freq
+        train_one_epoch(
+            model,
+            data,
+            loss,
+            0,
+            optimizer,
+            args.scaler, 
+            scheduler,
+            args
+        )
+        
+    # check that models weights are similar (within some threshold)
+        for p1, p2 in zip(model_accum_grad.parameters(), model_no_accum_grad.parameters()):
+            assert torch.allclose(p1, p2, atol=threshold)
+        
+
+def test_grad_acc_fsdp(accum_freq = 2, threshold = 1e-7):
+    args = MockArgs("open_lm_11m")
+    args.fsdp = True
```

Comment:
I think we want to test with: `args.fsdp = True` and `args.fsdp_amp = True`. Also this function basically same as the other one? so maybe factor out the common stuff?